storage/scan_queue.go

```diff
@@ -0,0 +1,145 @@
+//
+// The shouldQueue function combines the need for all three tasks into
+// a single priority. If any task is overdue, shouldQueue returns true.
+type scanQueue struct {
+	*baseQueue
+}
+
+// newScanQueue returns a new instance of scanQueue.
+func newScanQueue() *scanQueue {
+	sq := &scanQueue{}
+	sq.baseQueue = newBaseQueue(sq.shouldQueue, sq.process, scanQueueMaxSize)
+	return sq
+}
+
+// shouldQueue determines whether a range should be queued for
+// scanning, and if so, at what priority. Returns true for shouldQ in
+// the event that there are GC'able bytes, or it's been longer since
+// the last scan than the intent sweep or verification
+// intervals. Priority is derived from the addition of priority from
+// GC'able bytes and how many multiples of intent or verification
+// intervals have elapsed since the last scan.
+func (sq *scanQueue) shouldQueue(now time.Time, rng *Range) (shouldQ bool, priority float64) {
+	scanMeta, err := rng.GetScanMetadata()
+	if err != nil {
+		log.Errorf("unable to fetch scan metadata: %s", err)
+		return
+	}
+	elapsedNanos := now.UnixNano() - scanMeta.LastScanNanos
+
+	// Compute non-live bytes.
+	bytes, err := engine.GetRangeSize(rng.rm.Engine(), rng.Desc.RaftID)
+	if err != nil {
+		log.Errorf("unable to fetch range size stats: %s", err)
+		return
+	}
+	liveBytes, err := engine.GetRangeStat(rng.rm.Engine(), rng.Desc.RaftID, engine.StatLiveBytes)
+	if err != nil {
+		return
+	}
+	nonLiveBytes := bytes - liveBytes
+
+	// GC score.
+	estGCBytes := scanMeta.GC.EstimatedBytes(elapsedNanos, nonLiveBytes)
+	gcScore := float64(estGCBytes) / float64(gcByteCountNormalization)
+
+	// Intent sweep score. First check for intents. We only compute an
+	// intent score if there are any outstanding intents.
+	intentBytes, err := engine.GetRangeStat(rng.rm.Engine(), rng.Desc.RaftID, engine.StatIntentBytes)
+	if err != nil {
+		return
+	}
+	intentScore := float64(0)
+	if intentBytes > 0 {
+		intentScore = float64(elapsedNanos) / float64(intentSweepInterval.Nanoseconds())
+	}
+
+	// Verify score.
+	verifyScore := float64(elapsedNanos) / float64(verificationInterval.Nanoseconds())
+
+	// Compute priority.
+	if gcScore > 0 {
+		priority += gcScore
+	}
+	if intentScore > 1 {
+		priority += (intentScore - 1)
+	}
+	if verifyScore > 1 {
+		priority += (verifyScore - 1)
+	}
+	shouldQ = priority > 0
+	return
+}
+
+// process iterates through all keys in a range, calling the garbage
+// collector for each key and associated set of values. GC'd keys are
+// batched into InternalGC calls. Extant intents are resolved if
+// intents are older than intentAgeThreshold. The very act of scanning
+// keys verifies on-disk checksums, as each block checksum is checked
+// on load.
+func (sq *scanQueue) process(now time.Time, rng *Range) error {
+	snap := rng.rm.Engine().NewSnapshot()
+	iter := newRangeDataIterator(rng, snap)
+	defer iter.Close()
+	defer snap.Stop()
+
+	for ; iter.Valid(); iter.Next() {
```

Comment:
No, it's the next stage of this CL, or next CL if this one gets reviewed quickly enough. Right now the process method is non-functional.

In order to do GC, we're going to call the storage/engine/gc.go code with all keys which have multiple MVCC values. The scan will build up a list of keys with timestamp thresholds at which to GC earlier versions and send those via a new InternalGC RPC.